### PR TITLE
Implement opt-in sharp screen-space reflections for the deferred renderer.

### PR DESCRIPTION
**Pull request by pcwalton**
This commit implements *screen-space reflections* (SSR), which approximate real-time reflections based on raymarching through the depth buffer and copying samples from the final rendered frame. Numerous variations and refinements to screen-space reflections exist in the literature. This patch foregoes all of them in favor of implementing the bare minimum, so as to provide a flexible base on which to customize and build in the future.

For a general basic overview of screen-space reflections, see [1]. The raymarching shader uses the basic algorithm of tracing forward in large steps (what I call a *major* trace), and then refining that trace in smaller increments via binary search (what I call a *minor* trace). No filtering, whether temporal or spatial, is performed at all; for this reason, SSR currently only operates on very shiny surfaces. No acceleration via the hierarchical Z-buffer is implemented (though note that #12899 will add the infrastructure for this). Reflections are traced at full resolution, which is often considered slow. All of these improvements and more can be follow-ups.

SSR is built on top of the deferred renderer and is currently only supported in that mode. Forward screen-space reflections are possible albeit uncommon (though e.g. *Doom Eternal* uses them); however, they require tracing from the previous frame, which would add complexity. This patch leaves the door open to implementing SSR in the forward rendering path but doesn't itself have such an implementation. Screen-space reflections *are* supported in WebGL 2.

To add screen-space reflections to a camera, use the `ScreenSpaceReflectionsBundle` bundle or the `ScreenSpaceReflectionsSettings` component. In addition to `ScreenSpaceReflectionsSettings`, `DepthPrepass` and `DeferredPrepass` must also be present for the reflections to show up. The `ScreenSpaceReflectionsSettings` component contains several settings that artists can tweak, and also comes with sensible defaults.

A new example, `ssr`, has been added. It's loosely based on the [three.js ocean sample], but all the assets are original. Note that the three.js demo has no screen-space reflections and instead renders a mirror world.

Additionally, this patch fixes a random bug I ran across: that the `"TONEMAP_METHOD_ACES_FITTED"` `#define` is incorrectly supplied to the shader as `"TONEMAP_METHOD_ACES_FITTED "` (with an extra space) in some paths.

[1]: https://lettier.github.io/3d-game-shaders-for-beginners/screen-space-reflection.html

[three.js ocean sample]: https://threejs.org/examples/webgl_shaders_ocean.html

Screenshot 2024-04-13 174857

---

### Added

* Screen-space reflections can be enabled for very smooth surfaces by adding the `ScreenSpaceReflections` component to a camera. Deferred rendering must be enabled for the reflections to appear.

### CODE REVIEW

**Comment by pcwalton** [comment]
This should be ready for review now.

File: crates/bevy_pbr/src/ssr/mod.rs [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+/// When objects leave the camera, they will disappear from reflections.
+/// Alternatives that don't suffer from this problem include
+/// [`crate::environment_map::ReflectionProbeBundle`]s. The advantage of SSR is
+/// that it can reflect all objects, not just static ones.
+///
+/// SSR is an approximation technique and produces artifacts in some situations.
+/// Hand-tuning the settings in this component will likely be useful.
+#[derive(Clone, Copy, Component, Reflect, ShaderType)]
+#[reflect(Component)]
+pub struct ScreenSpaceReflections {
```
Comment (Kurble) [inline crates/bevy_pbr/src/ssr/mod.rs:72]: Isn't the convention that these kind of components are called `ScreenSpaceReflectionsSettings`?

Comment (pcwalton) [reply crates/bevy_pbr/src/ssr/mod.rs]: I did the renaming.

**Comment by pcwalton** [comment]
I renamed `ScreenSpaceReflections` to `ScreenSpaceReflectionsSettings` and added a bundle, `ScreenSpaceReflectionsBundle`.

File: Cargo.toml [hunk 2]
```diff
@@ -2871,6 +2871,16 @@ description = "Demonstrates FPS overlay"
 category = "Dev tools"
 wasm = true
 
+[[example]]
+name = "ssr"
+path = "examples/3d/ssr.rs"
```
Comment (mockersf) [inline Cargo.toml:2876]:
```suggestion
path = "examples/3d/ssr.rs"
doc-scrape-examples = true
```

**Comment by aevyrie** [comment]
Don't have time for a review, but saw the request. Just wanted to note that specular occlusion should use SSR if it is available. I believe I cited some papers in the shader code that talk about how to use SSR with specular occlusion. This might be out of scope, but it popped in my head when I saw this.

**Comment by pcwalton** [comment]
@aevyrie Yes, I agree. I think that should be a followup as it'll be a decent amount of code.

**Comment by chronicl** [comment]
The example crashes if any other prepasses are added to the camera, for example `NormalPrepass`. 
```
 Incompatible bind group at index 0 in the current render pipeline
      note: Should be compatible an with an explicit bind group layout with label = `mesh_view_layout_depth_deferred`
      note: Assigned explicit bind group layout with label = `mesh_view_layout_depth_normal_deferred`
 ```
I believe the `MeshPipelineViewLayoutKey` of the ssr pipeline should be adjusted for those cases, unless I'm doing something wrong or this is intended.

**Comment by pcwalton** [comment]
@chronicl Fixed, thanks!

File: crates/bevy_pbr/src/deferred/mod.rs [hunk 3]
```diff
@@ -473,6 +487,9 @@ pub fn prepare_deferred_lighting_pipelines(
         if ssao {
             view_key |= MeshPipelineKey::SCREEN_SPACE_AMBIENT_OCCLUSION;
         }
+        if ssr {
+            view_key |= MeshPipelineKey::SSR;
```
Comment (JMS55) [inline crates/bevy_pbr/src/deferred/mod.rs:491]: SSR -> SCREEN_SPACE_REFLECTIONS

File: crates/bevy_pbr/src/render/mesh_view_bindings.rs [hunk 4]
```diff
@@ -33,7 +36,9 @@ use crate::{
     },
     prepass, FogMeta, GlobalLightMeta, GpuFog, GpuLights, GpuPointLights, LightMeta,
     LightProbesBuffer, LightProbesUniform, MeshPipeline, MeshPipelineKey, RenderViewLightProbes,
-    ScreenSpaceAmbientOcclusionTextures, ShadowSamplers, ViewClusterBindings, ViewShadowBindings,
+    ScreenSpaceAmbientOcclusionTextures, ScreenSpaceReflectionsBuffer,
+    ScreenSpaceReflectionsSettings, ShadowSamplers, ViewClusterBindings, ViewShadowBindings,
+    CLUSTERED_FORWARD_STORAGE_BUFFER_COUNT, MESH_PIPELINE_VIEW_LAYOUT_SAFE_MAX_TEXTURES,
```
Comment (mockersf) [inline crates/bevy_pbr/src/render/mesh_view_bindings.rs:41]: `CLUSTERED_FORWARD_STORAGE_BUFFER_COUNT` is only available with `debug_assertions`, you must import it on its own behind the cfg check

Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_view_bindings.rs:41]: I'm confused. `CLUSTERED_FORWARD_STORAGE_BUFFER_COUNT` seems to be available in `bevy_pbr/src/render/light.rs` without a `#[cfg(debug_assertions)]` above it.

Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_view_bindings.rs:41]: I think you meant `MESH_PIPELINE_VIEW_LAYOUT_SAFE_MAX_TEXTURES`. I fixed this.

Comment (mockersf) [reply crates/bevy_pbr/src/render/mesh_view_bindings.rs]: right sorry 🤦

**Review by JMS55 — CHANGES_REQUESTED** [review]
Works surprisingly well with TAA, I was expecting the reflections to cause issues.

A bit surprised that the water reflection from the envmap is part of SSR, and not just part of the PBR shader, but maybe there's a good reason for that.

Some weird shader defs / things that stand out to me about how SSR is setup on the CPU. Not sure it's entirely right (although, it does work).

This PR is going to conflict with the clearcoat PR, probably want to merge that one first.

File: examples/3d/ssr.rs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        render_resource::{AsBindGroup, ShaderRef, ShaderType},
+        texture::{
+            ImageAddressMode, ImageFilterMode, ImageLoaderSettings, ImageSampler,
+            ImageSamplerDescriptor,
+        },
+    },
+};
+
+// The speed of camera movement.
+const CAMERA_KEYBOARD_ZOOM_SPEED: f32 = 0.1;
```
Comment (IceSentry) [inline examples/3d/ssr.rs:25]: You can do it if you feel like it but it's not required at all to merge. (I already made a note for me to do that in a follow-up PR). 

We generally try to put things like camera controllers and debug text in a separate submodule so only the setup part that shows how to enable the feature is at the top.

**Review by IceSentry — COMMENTED** [review]
Like mentioned in discord, there's currently a bug where the image gets brighter when toggling SSR. 

There also seems to be an issue with the transmission example. It's probably just a wrong ifdef.

Other than that, the code is high quality and everything looks good.

File: crates/bevy_pbr/src/deferred/mod.rs [hunk 6]
```diff
@@ -301,6 +305,10 @@ impl SpecializedRenderPipeline for DeferredLightingLayout {
             shader_defs.push("MOTION_VECTOR_PREPASS".into());
         }
 
+        if key.contains(MeshPipelineKey::SCREEN_SPACE_REFLECTIONS) {
+            shader_defs.push("SSR".into());
```
[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_pbr/src/deferred/mod.rs:309]: Yeah, good point, that would be more consistent with SSAO

**Comment by IceSentry** [comment]
@JMS55 

> A bit surprised that the water reflection from the envmap is part of SSR, and not just part of the PBR shader, but maybe there's a good reason for that.

Unless I'm misunderstanding what you are saying. The envmap reflection is still handled by the PBR shader when SSR is not present and when it's present it's still done when the roughness is under a certain threshold.

File: crates/bevy_pbr/src/ssr/mod.rs [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+        entity,
+        extracted_view,
+        has_environment_maps,
+        has_normal_prepass,
+        has_motion_vector_prepass,
+    ) in &views
+    {
+        // SSR is only supported in the deferred pipeline, which has no MSAA
+        // support. Thus we can assume MSAA is off.
+        let mut mesh_pipeline_view_key = MeshPipelineViewLayoutKey::from(Msaa::Off)
```
[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_pbr/src/ssr/mod.rs:331]: The depth prepass is also needed, otherwise depth isn't generated by the deferred pass

**Comment by JMS55** [comment]
In the SSR example, disabling SSR turns off both the cube reflection, but also the reflection on the water from the env map, which surprises me.

**Comment by pcwalton** [comment]
The reason why SSR delays evaluation of the environment map until the SSR pass is that the specular part of the environment map is only taken into account if the ray misses (after all, the environment map is "infinitely" far away), and we don't know whether the ray is going to miss until we evaluate SSR.

In other words, SSR effectively doubles as raymarched occlusion for the environment map.

**Comment by JMS55** [comment]
Oh I guess doesn't have to be this PR, but it's probably better to use a compute shader for SSR.

**Comment by IceSentry** [comment]
Yeah, definitely not this PR. I'd much rather have it as-is and introduce a compute path later. We'll also want to reuse the depth pyramid once we have that.

**Comment by pcwalton** [comment]
Will have to redo this later.

**Comment by pcwalton** [comment]
Closing as I am burned out and have no motivation to get this landed anymore.

**Pull request closed without merging**